docs/frontend-system: add app migration docs (#22391)
backstage/backstage: docs/frontend-system/building-apps/08-migrating.md

---
id: migrating
title: Migrating Apps
sidebar_label: Migration Guide
# prettier-ignore
description: How to migrate existing apps to the new frontend system
---

## Overview

This section describes how to migrate an existing Backstage app package to use the new frontend system. The app package is typically found at `packages/app` in your project and is responsible for wiring together the Backstage frontend application.

## Switching out `createApp`

The first step in migrating an app is to switch out the `createApp` function for the new one from `@backstage/frontend-api-app`:

```tsx title="in packages/app/src/App.tsx"
// highlight-remove-next-line
import { createApp } from '@backstage/app-defaults';
// highlight-add-next-line
import { createApp } from '@backstage/frontend-app-api';
```

This immediate switch will lead to a lot of breakages that we need to fix. Let's start by addressing `app.createRoot(...)`, which no longer accepts any arguments.

> Let's start by addressing the change to `app.createRoot(...)`, which no longer accepts any arguments. This represents a fundamental change that the new frontend system introduces. In the old system the app element tree that you passed to `app.createRoot(...)` was the primary way that you installed and configured plugins and features in your app. In the new system this is instead replaced by extensions that are wired together to an extension tree in the new system. Much more responsibility has been shifted to plugins in the new system, for example you no longer have to manually provide the route path for each plugin page, but instead only configure it if you want to override the default. For more information on how the new system works, see the [architecture](../architecture/01-index.md) section.

Reviewer: > Let's start by addressing `app.createRoot(...)`, which no longer accepts any arguments.

This sentence is repeated?

```bash
yarn add --cwd packages/app @backstage/core-compat-api
```

Once installed, import `convertLegacyApp`. If your app currently looks like this:

```tsx title="in packages/app/src/App.tsx"
const app = createApp({
  /* other options */
});

export default app.createRoot(
  <>
    <AlertDisplay transientTimeoutMs={2500} />
    <OAuthRequestDialog />
    <AppRouter>
      <Root>{routes}</Root>
    </AppRouter>
  </>,
);
```

Migrate it to the following:

```tsx title="in packages/app/src/App.tsx"
const legacyFeatures = convertLegacyApp(
  <>
    <AlertDisplay transientTimeoutMs={2500} />
    <OAuthRequestDialog />
    <AppRouter>
      <Root>{routes}</Root>
    </AppRouter>
  </>,
);

const app = createApp({
  /* other options */
  features: [...legacyFeatures],
});

export default app.createRoot();
```

We've taken all the elements that were previously passed to `app.createRoot(...)`, and instead passed them to `convertLegacyApp(...)`. We then pass the features returned by `convertLegacyApp` and forward them to the `features` option of the new `createApp`.

> There is one more details that we need to deal with before moving on. The `app.createRoot()` function now returns a React element rather and a component, so we need to update our app `index.tsx` as follows:

Reviewer: ```suggestion
There is one more detail that we need to deal with before moving on. The `app.createRoot()` function now returns a React element rather and a component, so we need to update our app `index.tsx` as follows:
```

For example, assuming you have a `lightTheme` extension that you want to add to your app, you can use the following:

```ts
const app = createApp({
  features: [
    createExtensionOverrides({
      extensions: [lightTheme],
    }),
  ],
});
```

### `apis`

[Utility API](../utility-apis/01-index.md) factories are now installed as extensions instead. Pass the existing factory to `createApiExtension` and install it in the app. For more information, see the section on [configuring Utility APIs](../utility-apis/04-configuring.md).

For example, the following API configuration:

```ts
const app = createApp({
  apis: [
    createApiFactory({
      api: scmIntegrationsApiRef,
      deps: { configApi: configApiRef },
      factory: ({ configApi }) => ScmIntegrationsApi.fromConfig(configApi),
    }),
  ],
});
```

Can be converted to the following extension:

```ts

> const scmIntegrationsApi = createApiExtension({

Reviewer: It wasn't immediate obvious to me but do we simply pass these extension creator return values mentioned throughout this doc into `createApp` the same way we provide the `lightTheme` extension mentioned above?
```tsx
 const app = createApp({
  features: [
    createExtensionOverrides({
      extensions: [lightTheme, scmIntegrationsApi, signInPage, ...],
    }),
  ],
});
```

Author: Yep! I added in an extra sentence after the `lightTheme` example to try to clarify it a bit more. I did want to try to avoid having to repeat the `createExtensionOverrides` throughout the entire doc, especially consider that you probably wanna group things together too 😅

  ),
});
```

### `configLoader`

The config loader API has been slightly changed. Rather than returning a promise for an array of `AppConfig` objects, it should now return the `ConfigApi` directly.

```ts
const app = createApp({
  async configLoader() {
    const appConfigs = await loadAppConfigs();
    // highlight-remove-next-line
    return appConfigs;
    // highlight-add-next-line
    return { config: ConfigReader.fromConfigs(appConfigs) };
  },
});
```

### `bindRoutes`

Route bindings can still be done using this option, but you now also have the ability to bind routes using static configuration instead. See the section on [binding routes](../architecture/07-routes.md#binding-external-route-references) for more information.

Note that if you are binding routes from a legacy plugin that was converted using `convertLegacyApp`, you will need to use the `convertLegacyRouteRefs` and/or `convertLegacyRouteRef` to convert the routes to be compatible with the new system.

For example, if both the `catalogPlugin` and `scaffolderPlugin` are legacy plugins, you can bind their routes like this:

```ts
const app = createApp({
  features: convertLegacyApp(...),
  bindRoutes({ bind }) {

>     bind(convertLegacyRouteRefs(catalogPlugin.createComponent), {

Reviewer: Should this still be accessed through the `externalRoutes` prop or did something change?
```suggestion
    bind(convertLegacyRouteRefs(catalogPlugin.externalRoutes.createComponent), {
```

Author: Ah this was a copy-paste derp, should be just `.externalRoutes` here and `createComponent` instead of `registerApi` below. Fix'd

Given that the app element tree is most of what builds up the app, it's likely also going to be the majority of the migration effort. In order to make the migration as smooth as possible we have provided a helper that lets you convert an existing app element tree into plugins that you can install in a new app. This in turn allows for a gradual migration of individual plugins, rather than needing to migrate the entire app structure at once.

The helper is called `convertLegacyApp` and is exported from the `@backstage/core-compat-api` package, which you will need to add as a dependency to your app package:

```bash

> yarn add --cwd packages/app @backstage/core-compat-api

Reviewer 2: ```suggestion
yarn --cwd packages/app add @backstage/core-compat-api
```

The `Router` component is now a built-in extension that you can override using `createRouterExtension`.

The Sign-in page is now installed as an extension using the `createSignInPageExtension` instead. See the section on [configuring the sign-in page](./index.md#TODO) for more information.

For example, the following sign-in page configuration:

```tsx
const app = createApp({
  components: {
    SignInPage: props => (
      <SignInPage
        {...props}
        provider={{
          id: 'github-auth-provider',
          title: 'GitHub',
          message: 'Sign in using GitHub',
          apiRef: githubAuthApiRef,
        }}
      />
    ),
  },
});
```

Can be converted to the following extension:

```tsx
const signInPage = createSignInPageExtension({
  loader: async () => props =>
    (
      <SignInPage
        {...props}
        provider={{
          id: 'github-auth-provider',
          title: 'GitHub',
          message: 'Sign in using GitHub',
          apiRef: githubAuthApiRef,
        }}
      />
    ),
});
```

### `themes`

> Themes are now installed as extensions instead using `createThemeExtension`. See the section on [configuring themes](./index.md#TODO) for more information.

Reviewer 2: ```suggestion
Themes are now installed as extensions, using `createThemeExtension`. See the section on [configuring themes](./index.md#TODO) for more information.
```

For example, the following theme configuration:

```tsx
const app = createApp({
  themes: [
    {
      id: 'light',
      title: 'Light',
      variant: 'light',
      Provider: ({ children }) => (
        <UnifiedThemeProvider theme={customLightTheme} children={children} />
      ),
    },
  ],
```

Can be converted to the following extension:

```tsx
const lightTheme = createThemeExtension({
  id: 'light',
  title: 'Light Theme',
  variant: 'light',
  icon: <LightIcon />,
  Provider: ({ children }) => (

>     <UnifiedThemeProvider theme={builtinThemes.light} children={children} />

Reviewer 2: ```suggestion
    <UnifiedThemeProvider theme={builtinThemes.light}>
      {children}
    </UnifiedThemeProvider>
```

it's kinda a bit clearer and i also see some lint rules and best practices avoiding a `children` prop. 🤷

```ts
const app = createApp({
  async configLoader() {
    const appConfigs = await loadAppConfigs();
    // highlight-remove-next-line
    return appConfigs;
    // highlight-add-next-line
    return { config: ConfigReader.fromConfigs(appConfigs) };
  },
});
```

### `bindRoutes`

Route bindings can still be done using this option, but you now also have the ability to bind routes using static configuration instead. See the section on [binding routes](../architecture/07-routes.md#binding-external-route-references) for more information.

Note that if you are binding routes from a legacy plugin that was converted using `convertLegacyApp`, you will need to use the `convertLegacyRouteRefs` and/or `convertLegacyRouteRef` to convert the routes to be compatible with the new system.

For example, if both the `catalogPlugin` and `scaffolderPlugin` are legacy plugins, you can bind their routes like this:

```ts
const app = createApp({
  features: convertLegacyApp(...),
  bindRoutes({ bind }) {
    bind(convertLegacyRouteRefs(catalogPlugin.externalRoutes), {
      createComponent: convertLegacyRouteRef(scaffolderPlugin.routes.root),
    });
  },
});
```

### `__experimentalTranslations`

> Translations are now installed as extensions instead using `createTranslationExtension`. See the section on [configuring translations](./index.md#TODO) for more information.

Reviewer 2: ```suggestion
Translations are now installed as extensions, using `createTranslationExtension`. See the section on [configuring translations](./index.md#TODO) for more information.
```

## Gradual Migration

> After updating all `createApp` options as well as using `convertLegacyApp` to use our existing app structure, you should be able to start up the app and see that it still works. If that is not the case, make sure you read any errors messages that you may see in the app as they can provide hints on what you need to fix. If you are still stuck, you can check if anyone else ran into the same issue in our [GitHub issues](https://github.com/backstage/backstage/issues), or ask for help in our [community Discord](https://discord.gg/backstage-687207715902193673).

Reviewer 2: ```suggestion
After updating all `createApp` options as well as using `convertLegacyApp` to use your existing app structure, you should be able to start up the app and see that it still works. If that is not the case, make sure you read any error messages that you may see in the app as they can provide hints on what you need to fix. If you are still stuck, you can check if anyone else ran into the same issue in our [GitHub issues](https://github.com/backstage/backstage/issues), or ask for help in our [community Discord](https://discord.gg/backstage-687207715902193673).
```

Once these migrations are complete you should be left with an empty `convertLegacyApp(...)` call that you can now remove, and your app should be fully migrated to the new system! 🎉

### Top-level Routes

Your top-level routes are the routes directly under the `AppRouter` component with the `<FlatRoutes>` element. In a small app they might look something like this:

```tsx title="in packages/app/src/App.tsx"
const routes = (
  <FlatRoutes>
    <Route path="/catalog" element={<CatalogIndexPage />} />
    <Route
      path="/catalog/:namespace/:kind/:name"
      element={<CatalogEntityPage />}
    >
      {entityPage}
    </Route>
    <Route path="/create" element={<ScaffolderPage />} />
    <Route
      path="/tech-radar"
      element={<TechRadarPage width={1500} height={800} />}
    />
  </FlatRoutes>
);
```

> Each of these routes need to be migrated to the new system. You can do it as gradually as you want, with the only restriction being that **all routes from a single plugin must be migrated at once**. This is because plugins discovered from these legacy routes will override any plugin that are installed in your app. If you for example only migrate one of the two routes defined by a plugin, the other route will remain and still override any plugin with the same ID, and you're left with a partial and likely broken plugin.

Reviewer 2: ```suggestion
Each of these routes needs to be migrated to the new system. You can do it as gradually as you want, with the only restriction being that **all routes from a single plugin must be migrated at once**. This is because plugins discovered from these legacy routes will override any plugins that are installed in your app. If you for example only migrate one of the two routes defined by a plugin, the other route will remain and still override any plugin with the same ID, and you're left with a partial and likely broken plugin.
```

To migrate a route, you need to remove it from your list of routes and instead install the new version of the plugin in your app. Before doing this you should make sure that the plugin supports the new system. Let's remove the scaffolder route as an example:

```tsx title="in packages/app/src/App.tsx"
const routes = (
  <FlatRoutes>
    <Route path="/catalog" element={<CatalogIndexPage />} />
    <Route
      path="/catalog/:namespace/:kind/:name"
      element={<CatalogEntityPage />}
    >
      {entityPage}
    </Route>
    {/* highlight-remove-next-line */}
    <Route path="/create" element={<ScaffolderPage />} />
    <Route
      path="/tech-radar"
      element={<TechRadarPage width={1500} height={800} />}
    />
  </FlatRoutes>
);
```

> If you are using [app feature discovery](../architecture/02-app.md#feature-discovery) the installation step is simple, it's already done! The new version of the scaffolder plugin was already discovered and present in the app, it was simply disabled because the plugin created from the legacy route had higher priority. If you do not use feature discovery, you will instead need to manually installed the new scaffolder plugin in your app through the `features` option of `createApp`.

Reviewer 2: ```suggestion
If you are using [app feature discovery](../architecture/02-app.md#feature-discovery) the installation step is simple, it's already done! The new version of the scaffolder plugin was already discovered and present in the app, it was simply disabled because the plugin created from the legacy route had higher priority. If you do not use feature discovery, you will instead need to manually install the new scaffolder plugin in your app through the `features` option of `createApp`.
```

```tsx
const nav = createExtension({
  namespace: 'app',
  name: 'nav',
  attachTo: { id: 'app/layout', input: 'nav' },
  output: {
    element: coreExtensionData.reactElement,
  },
  factory({ inputs }) {
    return {
      element: (
        <Sidebar>
          {/* Sidebar contents from packages/app/src/components/Root.tsx go here */}
        </Sidebar>
      ),
    };
  },
});
```

### App Root Elements

App root elements are React elements that are rendered adjacent to your current `Root` component. For example, in this snippet `AlertDisplay`, `OAuthRequestDialog` and `VisitListener` are all app root elements:

```tsx
export default app.createRoot(
  <>
    {/* highlight-next-line */}
    <AlertDisplay transientTimeoutMs={2500} />
    {/* highlight-next-line */}
    <OAuthRequestDialog />
    <AppRouter>
      {/* highlight-next-line */}
      <VisitListener />
      <Root>{routes}</Root>
    </AppRouter>
  </>,
);
```

> The `AlertDisplay` and `OAuthRequestDialog` are already provided as built-in extensions, and so will `VisitListener`. But, if you have your own custom root elements you will need to migrate them be extensions that you install in the app instead. Use `createAppRootElementExtension` to create said extension and then install it in the app.

Reviewer 2: ```suggestion
The `AlertDisplay` and `OAuthRequestDialog` are already provided as built-in extensions, and so will `VisitListener`. But, if you have your own custom root elements you will need to migrate them to be extensions that you install in the app instead. Use `createAppRootElementExtension` to create said extension and then install it in the app.
```

Whether the element used to be rendered as a child of the `AppRouter` or not doesn't matter. All new root app elements will be rendered as a child of the app router.

### App Root Wrappers

App root wrappers are React elements that are rendered as a parent of the current `Root` elements. For example, in this snippet the `CustomAppBarrier` is an app root wrapper:

```tsx
export default app.createRoot(
  <>
    <AlertDisplay transientTimeoutMs={2500} />
    <OAuthRequestDialog />
    <AppRouter>
      {/* highlight-next-line */}
      <CustomAppBarrier>
        <Root>{routes}</Root>
        {/* highlight-next-line */}
      </CustomAppBarrier>
    </AppRouter>
  </>,
);
```

> Any app root wrapper needs to be migrated to be an extension instead, using `createAppRootWrapperExtension`. Note that if you have multiple wrappers they must be completely independent of each other, the order in which the appear in the React tree should not matter. If that is not the case then you should group them into a single wrapper.

Reviewer 2: ```suggestion
Any app root wrapper needs to be migrated to be an extension instead, using `createAppRootWrapperExtension`. Note that if you have multiple wrappers they must be completely independent of each other, i.e. the order in which they appear in the React tree should not matter. If that is not the case then you should group them into a single wrapper.
```

## Tools

> - App Visualizer

Reviewer 2: todo here?
